File: bfsf.py
#!/usr/local/bin/python3
import sys

def lsout(loopstack):
    return str(loopstack).replace(', ','_').replace('[','').replace(']','')

def main():
    if len(sys.argv)!=2:
        print(f"Usage: {sys.argv[0]} <file>")
        return
    print("    .section .text")
    print("    .globl _start")
    print("_start:")
    print("    movabs $_my_data,%rsi")

    f=open(sys.argv[1],"rt")
    loopstack=[]
    lf='['
    for s in f:
        for c in s:
            if c=='>':
                print("    inc     %rsi")
            elif c=='<':
                print("    dec     %rsi")
            elif c=='+':
                print("    incb     (%rsi)")
            elif c=='-':
                print("    decb     (%rsi)")
            elif c=='.':
                print("    mov     $4,%rax") # syscall Write
                print("    mov     $1,%rdi") # fd: stdout
                print("    mov     $1,%rdx") # count
                print("    syscall")
            elif c==',':
                print("    mov     $3,%rax") # syscall Read
                print("    mov     $0,%rdi") # fd: stdin
                print("    mov     $1,%rdx") # count
                print("    syscall")
            elif c=='[':
                if lf==']':
                    loopstack[len(loopstack)-1]+=1
                else:
                    loopstack.append(1)
                lf='['
                print(f"LB{lsout(loopstack)}:")
                print("    cmpb   $0,(%rsi)")
                print(f"    je     LE{lsout(loopstack)}")
            elif c==']':
                if lf==']':
                    loopstack.pop()
                lf=']'
                print(f"    jmp     LB{lsout(loopstack)}")
                print(f"LE{lsout(loopstack)}:")
    f.close()
    print("    mov     $1,%rax")  # syscall exit
    print("    xor     %rdi,%rdi") # status = 0
    print("    syscall")
    print(".section .bss")
    print("_my_data: .zero 65536")

if __name__=='__main__':
    main()
    exit(0)

File: bfsf.go
package main

import (
    "fmt"
    "os"
    "io"
)

var (
    lf int ='['
    loopstack=[]int{}
)

func lsout(loopstack [] int) {
    for i:=0;i<len(loopstack);i++ {
        fmt.Printf("%d_",loopstack[i])
    }
}

func main() {
    if len(os.Args)!=2 {
        fmt.Println("Usage: go run bfs.go file.bf >out.s")
        return
    }
    fmt.Println("\t.section .text");
    fmt.Println("\t.globl _start");
    fmt.Println("_start:");
    fmt.Println("\tmovabs\t$_my_data,%rsi");


    file,err := os.Open(os.Args[1])
    byteData,err:=io.ReadAll(file)
    _ =err

    for idx:=0;idx<len(byteData);idx++ {
        switch byteData[idx] {
        case '>':
            fmt.Println("\tinc\t%rsi")
        case '<':
            fmt.Println("\tdec\t%rsi")
        case '+':
            fmt.Println("\tincb\t(%rsi)")
        case '-':
            fmt.Println("\tdecb\t(%rsi)")
        case '.':
	    fmt.Println("\tmov\t$4,%rax")
     	    fmt.Println("\tmov\t$1,%rdi")
	    fmt.Println("\tmov\t$1,%rdx")
	    fmt.Println("\tsyscall")
        case ',':
            fmt.Println("\tmov\t$3,%rax")
	    fmt.Println("\tmov\t$1,%rdi")
	    fmt.Println("\tmov\t$1,%rdx")
	    fmt.Println("\tsyscall\n")
        case '[':
            if lf==']' {
                loopstack[len(loopstack)-1]+=1
            } else {
                loopstack = append(loopstack,1)
            }
            lf='['
            fmt.Print("LB")
            lsout(loopstack)
            fmt.Println(":")
	    fmt.Println("\tcmpb\t$0,(%rsi)")
            fmt.Print("\tje     LE")
            lsout(loopstack)
            fmt.Println("")
        case ']':
            if lf==']' {
                loopstack=loopstack[:len(loopstack)-1]
            }
            lf=']'
            fmt.Print("\tjmp\tLB")
            lsout(loopstack)
            fmt.Println("")
            fmt.Print("LE")
            lsout(loopstack)
            fmt.Println(":")
        default:
        }
    }

    fmt.Println("\tmov\t$1,%rax")
    fmt.Println("\txor\t%rdi,%rdi")
    fmt.Println("\tsyscall")
    fmt.Println(".section .bss")
    fmt.Println("_my_data: .zero 65536")
}

